File: src/leetcode/LeetCode094.java
package leetcode;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class LeetCode094 {
    /**
     * 94. 二叉树的中序遍历
     * 给定一个二叉树，返回它的中序 遍历。
     * 示例:
     * 输入: [1,null,2,3]
     *    1
     *     \
     *      2
     *     /
     *    3
     * 输出: [1,3,2]
     */

    public static void main(String[] args) {
        System.out.println();
    }

    public List<Integer> inorderTraversal(TreeNode root) {
        if (root == null) {
            return null;
        }
        List<Integer> list = new ArrayList<>();
        Stack<TreeNode> stack = new Stack<>();
        TreeNode cur = root;
        while (!stack.isEmpty() || cur != null) {
            if (cur != null) {
                stack.push(cur);
                cur = cur.left;
            }
            if (cur == null) {
                cur = stack.pop();
                list.add(cur.val);
                cur = cur.right;
            }
        }
        return list;
    }

    private static class TreeNode {
        int val;
        TreeNode left;
        TreeNode right;
        TreeNode(int x) {
            val = x;
        }
    }
}

File: src/leetcode/LeetCode094.kt
package leetcode

import java.util.*

/*
94. 二叉树的中序遍历
给定一个二叉树，返回它的中序 遍历。
示例:
输入: [1,null,2,3]
   1
    \
     2
    /
   3
输出: [1,3,2]
进阶: 递归算法很简单，你可以通过迭代算法完成吗？
 */

fun inorderTraversal(root: TreeNode?): List<Int> {
    var list = mutableListOf<Int>()
    if (root == null) {
        return list
    }
    var stack = Stack<TreeNode>()
    var curNode: TreeNode? = root
    while (curNode != null || !stack.empty()) {
        while (curNode != null) {
            stack.push(curNode)
            curNode = curNode.left
        }
        curNode = stack.pop()
        list.add(curNode.`val`)
        curNode = curNode.right
    }
    return list
}